cpp/src/arrow/dataset/scan_node.cc

```diff
@@ -0,0 +1,653 @@
+  }
+
+  Status StartProducing() override {
+    START_COMPUTE_SPAN(span_, std::string(kind_name()) + ":" + label(),
+                       {{"node.kind", kind_name()},
+                        {"node.label", label()},
+                        {"node.output_schema", output_schema()->ToString()},
+                        {"node.detail", ToString()}});
+    END_SPAN_ON_FUTURE_COMPLETION(span_, finished_);
+    // std::cout << "StartProducing\n";
+    ARROW_ASSIGN_OR_RAISE(Future<> list_task, plan_->BeginExternalTask());
+    if (!list_task.is_valid()) {
+      // Cancelled before we even started
+      return Status::OK();
+    }
+    // std::cout << "LIST TASK: BEGIN\n";
+
+    AsyncGenerator<std::shared_ptr<Fragment>> frag_gen =
+        GetFragments(options_.dataset.get(), options_.filter);
+    Future<> visit_task = VisitAsyncGenerator(
+        std::move(frag_gen), [this](const std::shared_ptr<Fragment>& fragment) {
+          // std::cout << "Queuing fragment: " + fragment->ToString() + "\n";
+          QueueOrStartFragment(fragment);
+          return Status::OK();
+        });
+    visit_task
+        .Then(
+            [this]() {
+              // std::cout << "Marking listing completed\n";
+              if (synchronization_.MarkListingCompleted()) {
+                // It's possible, but probably unlikely, that all fragment scans have
+                // already finished and so we need to report the total batch count here in
+                // this case
+                batch_output_->InputFinished();
+                finished_.MarkFinished();
+              }
+            },
+            [this](const Status& err) {
+              synchronization_.MarkListingFailed(err);
+              outputs_[0]->ErrorReceived(this, err);
+            })
+        .AddCallback([list_task](const Status& st) mutable {
+          // std::cout << "LIST TASK: END\n";
+          list_task.MarkFinished();
+        });
+
+    return Status::OK();
+  }
+
+  void PauseProducing(ExecNode* output, int32_t counter) override {
+    // FIXME(TODO)
+    // Need to ressurect AsyncToggle and then all fragment scanners
+    // should share the same toggle
+  }
+
+  void ResumeProducing(ExecNode* output, int32_t counter) override {
+    // FIXME(TODO)
+  }
+
+  void StopProducing(ExecNode* output) override {
+    DCHECK_EQ(output, outputs_[0]);
+    StopProducing();
+  }
+
+  void StopProducing() override { synchronization_.Cancel(); }
+
+ private:
+  struct FragmentScanTask;
+  using ScanTaskPtr = std::list<FragmentScanTask>::iterator;
+  struct FragmentScanTask {
+    std::unique_ptr<FragmentEvolutionStrategy> fragment_evolution;
+    std::shared_ptr<FragmentScanner> fragment_scanner;
+    uint64_t unused_readahead_bytes;
+    bool iterating_tasks = false;
+    // Transitions to true when we've finished listing fragments for this task
+    bool finished = false;
+    // Transitions to true when we receieve an error or cancellation is requested
```

Comment:
Might it make sense to have an enum instead of individual flags?

Reply:
This code went away (and the complexity went into the async task scheduler for the most part)